docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,504 @@
+if and how they have already tested those algorithms. In many cases, it's
+acceptable to test them in just one place (and maybe through a different API
+entirely), and rely only on surface-level testing everywhere else. While it's
+always possible for more tests to uncover new bugs, the chances may be slim.
+The time we spend writing tests is highly valuable, so we have to be efficient!
+
+*Example* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4)
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+>    [[SELECTORS4]](https://dom.spec.whatwg.org/#biblio-selectors4).
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
+
+Many algorithms operate on input that takes the form of a scalar value. When
+the set of values is finite, it can be tempting to test them all exhaustively.
+When the set is very large, test authors can reduce repetition by defining
+tests programmatically in loops.
+
+Using advanced control flow techniques to dynamically generate tests has a
+number of drawbacks. It makes the test suite more difficult to understand since
+readers have to mentally "unwind" the iteration to determine what is actually
+being verified. It also increases the risk of bugs in the tests. These bugs may
+not be obvious--they may not cause failures, and they may exercise fewer cases
+than intended.
+
+Generally speaking, such exhaustive approaches are unlikely to catch more bugs
+than a handful of carefully-chosen test cases. Although the risks of dynamic
+test generation may be tolerable in some specific cases, it's usually best to
```

Comment:
This is too strong of a rebuke of generated tests for me to accept :)

Reply:
Despite my above dislike of generated tests, I agree with Philip. Dynamically generating `i = 1 to 10,000` is bad, but the sort of dynamic generation seen in `web-animations` is covering significantly more than 'a handful of carefully-chosen test cases' imo.